oximeter/impl/src/schema/ir.rs

```diff
@@ -0,0 +1,1288 @@
+                                &metric_field_names,
+                            )?;
+                            let authz_scope = extract_authz_scope(
+                                metric_name,
+                                self.target.authz_scope,
+                                &field_schema,
+                            )?;
+                            let timeseries_name = TimeseriesName::try_from(
+                                format!("{}:{}", target_name, metric_name),
+                            )?;
+                            let version = NonZeroU8::new(*version).unwrap();
+                            let description = TimeseriesDescription {
+                                target: self.target.description.clone(),
+                                metric: metric.description.clone(),
+                            };
+                            let schema = TimeseriesSchema {
+                                timeseries_name: timeseries_name.clone(),
+                                description,
+                                field_schema,
+                                datum_type: metric.datum_type,
+                                version,
+                                authz_scope,
+                                units: metric.units,
+                                created: Utc::now(),
+                            };
+                            if let Some(old) = timeseries
+                                .insert((timeseries_name, version), schema)
+                            {
+                                return defn_error(format!(
+                                    "Timeseries '{}' version {} is duplicated",
+                                    old.timeseries_name, old.version,
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(timeseries.into_values().collect())
+    }
+}
+
+#[derive(Clone, Debug)]
+enum CurrentVersion {
+    Active { version: NonZeroU8, fields: BTreeSet<String> },
+    Inactive { removed_in: NonZeroU8 },
+}
+
+impl CurrentVersion {
+    fn version(&self) -> NonZeroU8 {
+        match self {
+            CurrentVersion::Active { version, .. } => *version,
+            CurrentVersion::Inactive { removed_in } => *removed_in,
+        }
+    }
+
+    fn fields(&self) -> Option<&BTreeSet<String>> {
+        match self {
+            CurrentVersion::Active { fields, .. } => Some(fields),
+            CurrentVersion::Inactive { .. } => None,
+        }
+    }
+}
+
+fn defn_error<T>(s: String) -> Result<T, MetricsError> {
+    Err(MetricsError::SchemaDefinition(s))
+}
+
+/// Load the list of timeseries schema from a schema definition in TOML format.
+pub fn load_schema(
+    contents: &str,
+) -> Result<Vec<TimeseriesSchema>, MetricsError> {
+    toml::from_str::<TimeseriesDefinition>(contents)
+        .map_err(|e| MetricsError::Toml(e.to_string()))
```

Comment:
Tiny nit - can we preserve the full error type here, instead of squashing it into `e.to_string()`? If it's important to keep it a string, maybe consider using `InlineErrorChain::new(&e).to_string()` so we get any upstream errors included?

Reply:
IIRC, I tried keeping the error directly but hit some trait bound. I'm not sure what that would have been though, now that I see the error implements the common traits like `Clone` and `Display`. I'll try it again, and possibly use the inline-error chain trick if that doesn't work.

Reply:
Ah, this type shows up in our API, so it needs serialization and `JsonSchema` traits. I'll use the error-chain approach you recommend, thanks.

Reply:
Done in a410171f4